### ISSUE TITLE
Permission denied on mkdir /bitnami/mariadb with boot2docker

### ISSUE THREAD

**Issue by insuusvenerati** [bitnami/bitnami-docker-wordpress#110]
**Description**

When spinning up a bare docker-machine -d virtualbox and running docker-compose up after changing environment the mariadb container exits with error `mariadb_1    | Error executing 'postInstallation': EACCES: permission denied, mkdir '/bitnami/mariadb'`

**Steps to reproduce the issue:**

1. docker-machine create -d virtualbox <name>
2. docker-machine env <name>
3. docker-compose up

**Describe the results you received:**

https://h.stiforr.com/awawusiqow.coffeescript

**Describe the results you expected:**

mariadb container start.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Version**

```
- Output of `docker version`:
Client:
 Version:	18.01.0-ce
 API version:	1.35
 Go version:	go1.9.2
 Git commit:	03596f51b1
 Built:	Sun Jan 14 23:10:39 2018
 OS/Arch:	linux/amd64
 Experimental:	false
 Orchestrator:	swarm

Server:
 Engine:
  Version:	18.01.0-ce
  API version:	1.35 (minimum version 1.12)
  Go version:	go1.9.2
  Git commit:	03596f5
  Built:	Wed Jan 10 20:13:12 2018
  OS/Arch:	linux/amd64
  Experimental:	false
```

- Output of `docker info`:

```
Containers: 2
 Running: 0
 Paused: 0
 Stopped: 2
Images: 3
Server Version: 18.01.0-ce
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 29
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 89623f28b87a6004d4b785663257362d1658a729
runc version: b2567b37d7b75eb4cf325b77297b140ea686ce8f
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.4.111-boot2docker
Operating System: Boot2Docker 18.01.0-ce (TCL 8.2.1); HEAD : 0bb7bbd - Thu Jan 11 16:32:39 UTC 2018
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 995.9MiB
Name: wordpress
ID: ZRIK:TC42:IOQ7:K3AP:GH37:OFQD:KGRF:2LL6:5MTL:5ORY:X4XO:HZX6
Docker Root Dir: /mnt/sda1/var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 19
 Goroutines: 34
 System Time: 2018-01-29T22:26:17.682614196Z
 EventsListeners: 0
Username: stiforr
Registry: https://index.docker.io/v1/
Labels:
 provider=virtualbox
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

```

- Output of `docker-compose version` (if applicable):

```
docker-compose version 1.18.0, build unknown
docker-py version: 2.7.0
CPython version: 3.6.4
OpenSSL version: OpenSSL 1.1.0g  2 Nov 2017
```

**Additional environment details (AWS, VirtualBox, Docker for MAC, physical, etc.):**
```
λ docker-machine version
docker-machine version 0.13.0, build HEAD
```
```
λ uname -a
Linux arch 4.14.13-1-ck-piledriver #1 SMP PREEMPT Wed Jan 10 16:00:16 EST 2018 x86_64 GNU/Linux
```

```
λ git status                                                     
On branch master
Your branch is up to date with 'origin/master'.
```

**Comment by juan131** [comment]
Hi @insuusvenerati 

It seems that you're attaching a volume to the MariaDB container with no writing permissions for the user **mysql**. You can try to change the permissions of the local directory attached as a volume to the MariaDB container to avoid that permission error:

```
$ chown -R 1001:1001 /path/to/your/mariadb_data
```

**Comment by insuusvenerati** [comment]
The mkdir error is coming from inside the container. I'm using the unmodified bitnami wordpress docker compose which uses named volumes so there isn't any volumes on the host. Might be worth nothing this issue doesn't occur on Ubuntu 16.04

**Comment by sameersbn** [comment]
@insuusvenerati I followed your instructions and did not face any issues. 

For reference:

```
$ docker version
Client:
 Version:	18.02.0-ce-rc1
 API version:	1.35
 Go version:	go1.9.2
 Git commit:	5e1d90a
 Built:	Thu Jan 25 00:33:50 2018
 OS/Arch:	darwin/amd64
 Experimental:	true
 Orchestrator:	kubernetes
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

However the issue you are noticing is due to the non-root changes done to the mariadb image. You should try either of the following docker-compose files to resolve this issue

**Option 1**

Launch the mariadb container as the `root` user. This means the container will have all the required permissions to give ownership of the mysql data directories to the `mysql` user.

```yaml
version: '2'
services:
  mariadb:
    user: root
    image: 'bitnami/mariadb:latest'
    volumes:
      - 'mariadb_data:/bitnami'
    environment:
      - MARIADB_USER=bn_wordpress
      - MARIADB_DATABASE=bitnami_wordpress
      - ALLOW_EMPTY_PASSWORD=yes
  wordpress:
    image: 'bitnami/wordpress:latest'
    labels:
      kompose.service.type: nodeport
    ports:
      - '80:80'
      - '443:443'
    volumes:
      - 'wordpress_data:/bitnami'
    depends_on:
      - mariadb
    environment:
      - MARIADB_HOST=mariadb
      - MARIADB_PORT_NUMBER=3306
      - WORDPRESS_DATABASE_USER=bn_wordpress
      - WORDPRESS_DATABASE_NAME=bitnami_wordpress
      - ALLOW_EMPTY_PASSWORD=yes
volumes:
  mariadb_data:
    driver: local
  wordpress_data:
    driver: local
```

**Option 2**

Create a container to fix the permissions of the mounted volume

```yaml
version: '2'
services:
  fix-mariadb-volume-ownership:
    image: 'bitnami/mariadb:latest'
    user: root
    command: chown -R 1001:1001 /bitnami
  mariadb:
    image: 'bitnami/mariadb:latest'
    volumes:
      - 'mariadb_data:/bitnami'
    environment:
      - MARIADB_USER=bn_wordpress
      - MARIADB_DATABASE=bitnami_wordpress
      - ALLOW_EMPTY_PASSWORD=yes
    depends_on:
      - fix-mariadb-volume-ownership
  wordpress:
    image: 'bitnami/wordpress:latest'
    labels:
      kompose.service.type: nodeport
    ports:
      - '80:80'
      - '443:443'
    volumes:
      - 'wordpress_data:/bitnami'
    depends_on:
      - mariadb
    environment:
      - MARIADB_HOST=mariadb
      - MARIADB_PORT_NUMBER=3306
      - WORDPRESS_DATABASE_USER=bn_wordpress
      - WORDPRESS_DATABASE_NAME=bitnami_wordpress
      - ALLOW_EMPTY_PASSWORD=yes
volumes:
  mariadb_data:
    driver: local
  wordpress_data:
    driver: local
```

**Comment by aleksander351** [comment]
The `fix-mariadb-volume-ownership` does not work for me.

I first create the mariadb_data and wordpress_data volumes before I run docker-compose up. But then it creates 2 more volumes: mysite_mariadb_data and mysite_wordpress_data.

**Comment by badmark** [comment]
This is not a solution, running as root is NEVER a good solution. Bitnami, you make bad Docker images and should be ashamed...

**Comment by juan131** [comment]
Hi @badmark 

> This is not a solution, running as root is NEVER a good solution

Most of our infrastructure images (such as MariaDB) are non-root by default (despite they can be run as a root container too). When running the container as root, please note that it doesn't mean that MariaDB process is running as root, we use a non-privileged user.

When dealing with this kind of errors (mounted volumes with bad permissions/ownership), there are different alternatives depending on the platform you're using to run your containers. For instance, on K8s you can use "securityContext" to address this issue, or use an "initContainer" with root privileges. 

We use the 1st approach by default in our Helm chart for MariaDB (this way we avoid running containers as root) as you can check in the link below:

- https://github.com/helm/charts/blob/master/stable/mariadb/values.yaml#L100

> Bitnami, you make bad Docker images and should be ashamed...

We're always open for including our users suggestions. So please feel free to provide any feedback you may have so we can improve our images.

**Comment by badmark** [comment]
Yeah, suggestion is not to use your images. Thought this would save me time, it only wasted it.

I can run the official MariaDB image without permission issues, but for Bitnami I'm expected to jump through hoops and do "non-standard" methods to provide permission within the container?

No thanks.

**Comment by juan131** [comment]
Hi @badmark 

I am glad that you found another solution in the market that met your expectations.

Please note we are not using GitHub as a "review platform", but for sharing the code and images we developed and provide support to their users. 

Users are also allowed to provide feedback so we can improve the container images, but comments with  no other purpose than fault-finding are not allowed and will be discarded.

**Comment by jmpolvera** [comment]
I made this in the host directory and worked:

sudo chown -R 1001:1001 <directory>


https://techoverflow.net/2018/12/15/solving-bitnami-docker-redmine-cannot-create-directory-bitnami-mariadb-permission-denied/

**Comment by juan131** [comment]
Hi @jmpolvera 

Thanks for sharing it! @sameersbn already mentioned this possibility in this comment (https://github.com/bitnami/bitnami-docker-wordpress/issues/110#issuecomment-361940514) where you can find how to automate it using a container.

**Comment by Franz333** [comment]
> I made this in the host directory and worked:
> 
> sudo chown -R 1001:1001
> 
> https://techoverflow.net/2018/12/15/solving-bitnami-docker-redmine-cannot-create-directory-bitnami-mariadb-permission-denied/

Hi @jmpolvera I saw this too and I applied this command in a host dir but then I could not modify files with my code editor..

Did you have the same problem?

**Comment by carrodher** [comment]
If the owner/group of the files changed in the host machine probably you will need to edit the files as a privileged user, it will depend on the user with whom you want to edit these files. Can you try editing them with something like `sudo vim`?

**Comment by Franz333** [comment]
Hi @carrodher,
Thanks for trying to help me. 
Here are two scenarios..

# No ownership/permissions change to host repositories
After cloning my distant repo containing the two directories that I will bind mount , here are the permissions:

Main folder: 
`drwxr-xr-x 4 franz franz 4096 Mar 26 10:45 uclockit-site`
Main folder > subfolders: 
`drwxr-xr-x 2 franz franz 4096 Mar 26 10:45 db`
`drwxr-xr-x 5 franz franz  4096 Mar 26 10:45 wp`

Result: 
```
Permission denied on mariab db container
mariadb 09:57:01.72 INFO ==> Initializing mariadb database
mkdir: cannot create directory '/bitnami/mariadb/data': Permission denied
```

Which is logical as I am running the image with bitnami user 1001 as advised in the doc

# Change ownership/permissions of host repositories to 1001
`sudo chown 1001:1001 uclockit-site -R`

Main folder:
`drwxr-xr-x 4 1001  1001  4096 Mar 26 10:45 uclockit-site`
Main folder > subfolders:
`drwxr-xr-x 3 1001 1001 4096 Mar 26 11:06 db`
`drwxr-xr-x 5 1001 1001 4096 Mar 26 10:45 wp`

No problem, wordpress is loading fine. Can remove plugin from wordpress admin dashboard etc...
But when I want to edit a file with visual studio code, I don't have the permission..

```
Failed to save 'style.css': Unable to write file 'vscode-remote://wsl+ubuntu-16.04/home/franz/uclockit/1002/bitnami-docker-wordpress/5/debian-10/uclockit-site/wp/wp-content/themes/astra-child/style.css' (NoPermissions (FileSystemError): Error: EACCES: permission denied, open '/home/franz/uclockit/1002/bitnami-docker-wordpress/5/debian-10/uclockit-site/wp/wp-content/themes/astra-child/style.css')
```

Which seems logical too..

So how do I set both permissions to user 1001 for the containers to start without any problems AND permissions to the user 'franz' who is working on the host files within the ubuntu wsl?

**Comment by carrodher** [comment]
You can try using the user and group as differentiated elements, I mean, the user `franz` and the group `1001` can be the owners of the `uclockit-site` directory, apart from that you need to ensure the user and the groups have the proper permission to write/read in that directory.

```console
$ ls -la mydir
total 16
drwxr-xr-x 4 bitnami bitnami 4096 Mar 29 11:32 .
drwxrwxr-x 6 bitnami bitnami 4096 Mar 29 11:31 ..
drwxr-xr-x 2 bitnami bitnami 4096 Mar 29 11:32 subdir1
drwxr-xr-x 2 bitnami bitnami 4096 Mar 29 11:32 subdir2
```
In the above example, `bitnami` is user and group at the same time, we can change the group by running
```console
$ sudo chgrp -R 1001 mydir

$ ls -la mydir
total 16
drwxr-xr-x 4 bitnami    1001 4096 Mar 29 11:32 .
drwxr-xr-x 2 bitnami    1001 4096 Mar 29 11:32 subdir1
drwxr-xr-x 2 bitnami    1001 4096 Mar 29 11:32 subdir2
```
In that case, now `bitnami` is the user and `1001` is the group, but taking a look at the permissions: `rwx r-x r-x` the first block reference the user, and the second one the group. In that case, the group block is missing the `w` (write) so we need to give write permission to the group:
```console
$ sudo chmod -R g+w mydir

$ ls -la mydir/
total 16
drwxrwxr-x 4 bitnami    1001 4096 Mar 29 11:32 .
drwxrwxr-x 2 bitnami    1001 4096 Mar 29 11:32 subdir1
drwxrwxr-x 2 bitnami    1001 4096 Mar 29 11:32 subdir2
```
now the `w` appears in the block corresponding to the group so this folder (and subfolders) should be readable and writable by the user `bitnami` and those users belonging to the `1001` group.

In your case, you should follow a similar approach in order to guarantee the correct permissions to the host user and 1001 user by playing with the above commands until reaching your desired status.

**Comment by Franz333** [comment]
Hi @carrodher,
**Awsome answer !** 
Thank you very much, it helped a lot. 
With your  explanation I think I managed to do something that's finally working.
As it can help other beginners like me, I will detail a bit. 
Because I am keeping bitnami default `1001` user, I changed the group of my wordpress host mounted files to assign it the group to which my "local user" belongs to. In my case, my user is `franz` and this user belongs to the group `franz`.
`sudo chown 1001:franz uclockit-site -R`
And as it was advised by @carrodher I added the "write" permission to that group:
`sudo chmod -R g+w uclockit-site`.
I can now run the wordpress container with my wordpress source files without permissions problems AND modify those files localy in my editor (child-theme files for example) without any problems.
Just mind that bitnami default installed plugins (and maybe new plugins installed from wp admin panel) will have different group ownership: `drwxrwxr-x  7 1001 root   4096 Mar 30 15:19 all-in-one-seo-pack`
Thanks again
